Add webhook receiver for MeB OAuth

amCap1712 opened the pull request.

amCap1712 committed:
Add webhook receiver for MeB OAuth

amCap1712 committed:
Use login_hint to improve user flow

login_hint=register takes users not logged in test.meb.org
to signup instead of signin page.

amCap1712 committed:
change last_login default and add webhook tests

amCap1712 commented:
todo:

1. remove /login page only showing sign in with musicbrainz on LB side.
2. update sign in link in navbar.

MonkeyDo commented on frontend/js/src/home/Homepage.tsx:
```diff
@@ -66,18 +67,24 @@ function HomePage() {
   );
   const loginButton = (
-    <Link className="login-button" to="/login/">
+    <a className="login-button" href={buildAuthUrl("login")}>
```
Making a note that #3398 also added a create account link in the navbar.

amCap1712 committed:
Implement webhook handlers

1. user.created - do not add email, will be handled by user.updated on email verification
2. user.updated - change in username or email (after verification of email is complete)
3. user.deleted - when user account is deleted in MeB.

amCap1712 committed:
Add buildAuthUrl util to consistently handle login redirects

Also, remove redundant /login page.

MonkeyDo reviewed (commented):
I've only properly looked at the front-end parts, but it looks fine at a glance.

How can I test this thoroughly? I'd like to click every login/logout link we have on the site...

MonkeyDo commented on frontend/js/src/login/Login.tsx:
```diff
@@ -1,73 +0,0 @@
-        <p className="text-danger" style={{ fontSize: "16pt" }}>
-          Important!
-        </p>
-        <p>
-          By signing into ListenBrainz, you grant the MetaBrainz Foundation
-          permission to include your listening history in data dumps we make
-          publicly available under the{" "}
-          <a href="https://creativecommons.org/publicdomain/zero/1.0/">
-            CC0 license
-          </a>
-          . None of your private information from your user profile will be
-          included in these data dumps.
-        </p>
-        <p>
-          Furthermore, you grant the MetaBrainz Foundation permission to process
-          your listening history and include it in new open source tools such as
-          recommendation engines that the ListenBrainz project is building. For
-          details on processing your listening history, please see our{" "}
-          <a href="https://metabrainz.org/gdpr">GDPR compliance statement</a>.
-        </p>
-        <p>
```
I feel like we should not entirely lose this, and should show this warning somewhere so users are informed. Not sure where would be appropriate though.
Maybe on the MEB website, if we know they are coming from LB, or if we have a generic message saying "your data in MetaBrainz projects will be publicly available"

amCap1712 committed:
Reconcile buildAuthUrl and getRegistrationUrl

amCap1712 committed:
fix test

amCap1712 committed:
rename metabrainz_webhook_secret to oauth/webhook_secret

amCap1712 committed:
replace new-oauth2 with oauth2

MonkeyDo committed:
Webhooks: expect field "name" instead of "username"

This is the format we receive from webhooks requests

amCap1712 committed:
Allow signups without email and gate listens/services

Accept new users without an email during login, update verified email from
user.updated webhooks, and reject listen submission/import APIs for users
without a verified email. Block new music service connections until email
verification while still allowing disconnects.

Centralize email verification messaging and MetaBrainz profile/blog links for
frontend settings pages.

MonkeyDo committed:
Merge branch 'master' into meb-oauth

MonkeyDo commented on listenbrainz/config.py.sample:
I suppose this should be changed to meb.org once we are ready to deploy?

MonkeyDo committed:
Fix webhooks tests

Tests expected wrong attribute. See d5ef95a2ea11ea009491c1b4d6a8f3f07903c407

MonkeyDo committed:
Remove mb_engine guard in no-email-user rejections conditions

Not needed anymore since we don't load user data from the musicbrainz database anymore

MonkeyDo committed:
Tests: Mock the OAuth URL based on the app config rather than hardcoded

MonkeyDo commented on listenbrainz/webserver/views/import_listens.py:
I think we don't need this `mb_engine` guard anymore (happens in a few files, next to `REJECT_LISTENS_WITHOUT_USER_EMAIL`), considering we don't use the MB database anymore for user account related stuff.

I think this might be causing the integration test failures related to rejection-without-email.

Also noting that we removed the mock previously in place to mock mb_engine [here](https://github.com/metabrainz/listenbrainz-server/blob/d42ecff4a7861119bc16811125d64cd7cc8116ef/listenbrainz/tests/integration/test_listens_importer.py#L120), which is likley why that test fails

Changed that in commit d9f3dce0cdad21421cf32975a7c25bcaefb713d8

MonkeyDo committed:
Fix indent option for orjson
